packages/rocketchat-lib/server/lib/notifyUsersOnMessage.js

```diff
@@ -1,10 +1,36 @@
 RocketChat.callbacks.add('afterSaveMessage', function(message, room) {
+
+	/**
+	 * Chechs if a messages contains a user highlight
+	 *
+	 * @param {string} message
+	 * @param {array|undefined} highlights
+	 *
+	 * @returns {boolean}
+     */
+	function messageContainsHighlight(message, highlights) {
+		if (! highlights || highlights.length == 0) { return false; }
+
+		var has = false;
+		highlights.forEach(function (highlight) {
+			if (message.msg.includes(highlight)) { has = true; }
+		});
+
+		return has;
+	}
+
+
 	if (room.t != null && room.t === 'd') {
 		// Update the other subscriptions
 		RocketChat.models.Subscriptions.incUnreadOfDirectForRoomIdExcludingUserId(message.rid, message.u._id, 1);
 	} else {
-		var mentionIds = [];
-		var toAll = false;
+		var mentionIds, toAll, highlightsIds, highlights;
+
+		mentionIds = [];
+		highlightsIds = [];
+		toAll = false;
+		highlights = RocketChat.models.Users.find({}, { fields: { '_id': 1, 'settings.preferences.highlights': 1 }}).fetch();
```

Comment:
we need to improve this query the most we can.. the main problems:
- this query will run on **every** message, so: 
  - we need to use a filter with an index.
  - we can't fetch() all users and loop all of them on every message sent
- we have to filter only users that can actually **see** the message (you don't want to notify someone not in the private group :smile: )